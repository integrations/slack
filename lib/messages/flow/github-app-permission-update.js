const ErrorMessage = require('./error-message');

module.exports = class GitHubAppPermissionUpdate extends ErrorMessage {
  constructor({ installation }) {
    super();
    this.installation = installation;
  }
  get guidance() {
    return '\n_Note: If you are not the owner of the repository or the owner of the organization, ' +
      'you will have to ask the owner to install it for you. ' +
    `Give them this link: ${this.installation.html_url}/permissions/update _`;
  }

  toJSON() {
    const message = this.getErrorMessage();
    message.attachments[0].text = `This feature requires updated permissions.\n${this.guidance}`;
    message.attachments[0].actions = [{
      style: 'primary',
      text: 'Update GitHub App permissions',
      type: 'button',
      url: `${this.installation.html_url}/permissions/update`,
    }];
    return message;
  }
};
