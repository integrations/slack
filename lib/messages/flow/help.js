const supportLink = require('../../support-link');

// FIXME: make this dynamic
const commands = [
  {
    usage: 'subscribe owner/repository',
    desc: 'Subscribe to notifications for a repository',
  },
  {
    usage: 'unsubscribe owner/repository',
    desc: 'Unsubscribe from notifications for a repository',
  },
  {
    usage: 'subscribe owner',
    desc: 'Subscribe to notifications for all repositories in an organization',
  },
  {
    usage: 'unsubscribe owner',
    desc: 'Unsubscribe from notifications for an organization',
  },
  {
    usage: 'subscribe list',
    desc: 'List all active subscriptions in a channel',
  },
  {
    usage: 'close [issue link]',
    desc: 'Close an issue',
  },
  {
    usage: 'reopen [issue link]',
    desc: 'Reopen an Issue',
  },
  {
    usage: 'settings',
    desc: 'Adjust your settings in this workspace',
  },
  {
    usage: 'signin',
    desc: 'Connect your GitHub account',
  },
  {
    usage: 'help',
    desc: 'Show this help message',
  },
  {
    usage: 'open owner/repository',
    desc: 'Create a new issue',
  },
];

if (process.env.DEPLOY_COMMAND_ENABLED) {
  commands.push({
    usage: 'deploy owner/repository',
    desc: 'Trigger a deployment',
  });

  commands.push({
    usage: 'deploy owner/repository list',
    desc: 'List deployments of a repo',
  });
}

const { Message } = require('../../messages');

module.exports = class Help extends Message {
  constructor(command, subcommand) {
    super({
      footer: [
        '<https://github.com/integrations/slack#readme|Learn More>',
        `<${supportLink()}|Contact Support>`,
      ].join(' â€” '),
    });
    this.command = command;
    this.subcommand = subcommand;
  }

  filterForSubCommand(command) {
    if (!this.subcommand) {
      return true;
    }
    return command.usage.includes(this.subcommand);
  }

  toJSON() {
    const attachments = commands
      .filter(command => this.filterForSubCommand(command))
      .map(command => ({
        text: `${command.desc}:\n\`${this.command} ${command.usage}\``,
        mrkdwn_in: ['text'],
      }));

    attachments.push({
      ...this.getBaseMessage(),
      text: '',
      color: '',
    });

    return {
      response_type: 'ephemeral',
      text: `:wave: Need some help with \`${this.command}\`?`,
      attachments,
    };
  }
};
