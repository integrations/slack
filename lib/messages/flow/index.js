/* eslint-disable global-require */
const Subscribed = require('./subscribed');
const InvalidUrl = require('./invalid-url');
const InstallGitHubApp = require('./install-github-app');
const GitHubAppPermissionUpdate = require('./github-app-permission-update');
const NotFound = require('./not-found');
const SignIn = require('./signin');
const Help = require('./help');
const AlreadySubscribed = require('./already-subscribed');
const NotSubscribed = require('./not-subscribed');
const ActivateLegacySubscriptions = require('./activate-legacy-subscriptions');
const ReEnableSubscription = require('./re-enable-subscription');
const Exception = require('./exception');
const SlackInstalled = require('./slack-installed');
const UpdatedSettings = require('./updated-settings');
const SignOut = require('./signout');

module.exports = {
  InvalidUrl,
  InstallGitHubApp,
  NotFound,
  Subscribed,
  SignIn,
  Help,
  AlreadySubscribed,
  NotSubscribed,
  ActivateLegacySubscriptions,
  ReEnableSubscription,
  Exception,
  SlackInstalled,
  UpdatedSettings,
  GitHubAppPermissionUpdate,
  SignOut,
};
