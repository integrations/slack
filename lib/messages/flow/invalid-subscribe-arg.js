const ErrorMessage = require('./error-message');

module.exports = class InvalidSubscribeArg extends ErrorMessage {
  constructor(subscribeInput) {
    super();
    this.subscribeInput = subscribeInput;
  }

  toJSON() {
    const message = this.getErrorMessage();
    message.attachments[0].text = `\`${this.subscribeInput}\` is not a valid argument.`;
    return message;
  }
};
