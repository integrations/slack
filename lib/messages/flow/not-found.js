const ErrorMessage = require('./error-message');

module.exports = class NotFound extends ErrorMessage {
  constructor(resource) {
    super();
    this.resource = resource;
  }

  toJSON() {
    const message = this.getErrorMessage();
    message.attachments[0].text = `Could not find resource: \`${this.resource}\``;
    return message;
  }
};
