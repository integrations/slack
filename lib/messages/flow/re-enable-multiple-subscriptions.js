const { Message } = require('..');

module.exports = class ReEnableMultipleSubscriptions extends Message {
  constructor(subscribedRepositories, subscribedAccounts, creator, reason) {
    super({});
    const total = subscribedRepositories.length + subscribedAccounts.length;
    this.plural = total !== 1;
    this.repositories = subscribedRepositories;
    this.accounts = subscribedAccounts;
    this.creator = creator;
    this.reason = reason;
  }

  get introduction() {
    const message = [
      this.repositories.length > 0 ? `${this.repositories.length} ${this.repositories.length > 1 ? 'repositories' : 'repository'}` : null,
      this.accounts.length > 0 ? `${this.accounts.length} ${this.accounts.length > 1 ? 'accounts' : 'account'}` : null,
    ].filter(Boolean).join(' and ');
    return `${this.plural ? 'Subscriptions' : 'The subscription'} to ${message} ${this.plural ? 'have' : 'has'} been disabled ` +
      `because <@${this.creator}>, who originally set ${this.plural ? 'them' : 'it'} up, has ${this.reason}.\n` +
      `Use the following slash command${this.plural ? 's' : ''} to re-enable the subscription${this.plural ? 's' : ''}:`;
  }

  get commands() {
    return [
      ...this.repositories.map(repository => (
        `/github subscribe ${repository.full_name}`
      )),
      ...this.accounts.map(account => (
        `/github subscribe ${account.login}`
      )),
    ].join('\n');
  }

  getAttachment() {
    return {
      ...this.getBaseMessage(),
      mrkdwn_in: ['text'],
      text: `${this.introduction}\n${this.commands}`,
    };
  }

  toJSON() {
    return {
      attachments: [this.getAttachment()],
      response_type: 'in_channel',
    };
  }
};
