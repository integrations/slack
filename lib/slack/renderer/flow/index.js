/* eslint-disable global-require */
const Subscribed = require('./subscribed');
const InvalidUrl = require('./invalid-url');
const InstallGitHubApp = require('./install-github-app');
const NotFound = require('./not-found');
const SignIn = require('./signin');
const Help = require('./help');
const AlreadySubscribed = require('./already-subscribed');
const NotSubscribed = require('./not-subscribed');
const ActivateLegacySubscriptions = require('./activate-legacy-subscriptions');

module.exports = {
  InvalidUrl,
  InstallGitHubApp,
  NotFound,
  Subscribed,
  SignIn,
  Help,
  AlreadySubscribed,
  NotSubscribed,
  ActivateLegacySubscriptions,
};
