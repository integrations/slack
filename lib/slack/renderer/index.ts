const constants = {
  GITHUB_BLACK: '#24292f',
  CLOSED_RED: '#cb2431',
  OPEN_GREEN: '#36a64f',
  MERGED_PURPLE: '#6f42c1',
  STATUS_SUCCESS: '#28a745',
  STATUS_PENDING: '#dbab09',
  STATUS_FAILURE: '#cb2431',
  BASE_ATTACHMENT_COLOR: '#24292f',
  ATTACHMENT_FIELD_LIMIT: 2,
  MAJOR_MESSAGES: {
    'pull_request.opened': true,
    'issues.opened': true,
  },
};

class Message {
  constructor(constructorObject) {
    this.includeFooter = constructorObject.includeFooter;
    this.footerURL = constructorObject.footerURL;
    this.includeAuthor = constructorObject.includeAuthor;
    this.author = constructorObject.author;
  }

  static cleanFields(
    fields,
    fieldLimit = constants.ATTACHMENT_FIELD_LIMIT,
    short = true,
  ) {
    return fields
      .filter(field => field.value)
      .map((field) => {
        if (short) {
          return { ...field, short: true };
        }
        return { ...field };
      })
      .slice(0, fieldLimit);
  }

  getBaseMessage() {
    const baseMessage = {
      color: constants.GITHUB_BLACK,
    };
    if (this.includeFooter) {
      baseMessage.footer = `<${this.footerURL}|View it on GitHub>`;
      baseMessage.footer_icon = 'https://assets-cdn.github.com/favicon.ico';
    }
    if (this.includeAuthor) {
      baseMessage.author_name = this.author.login;
      baseMessage.author_icon = this.author.avatarURL;
      baseMessage.author_link = this.author.htmlURL;
    }
    return baseMessage;
  }

}

// TODO: /github test-run -> delivers all webhooks we're currently ready to receive

module.exports = {
  Message,
  constants,
};
