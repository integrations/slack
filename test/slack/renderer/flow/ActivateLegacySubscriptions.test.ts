import { ActivateLegacySubscriptions } from "../../../../lib/slack/renderer/flow";

describe("AlreadySubscribed message rendering", () => {
  test("works for multiple configurations and authors in a channel", async () => {
    const subscriptions = [
      {
        authorSlackId: "U12345",
        repoFullName: "kubernetes/kubernetes",
      },
      {
        authorSlackId: "U12345",
        repoFullName: "facebook/react",
      },
      {
        authorSlackId: "U54321",
        repoFullName: "django/django",
      },
      {
        authorSlackId: "U54321",
        repoFullName: "microsoft/typescript",
      },
    ];
    const message = new ActivateLegacySubscriptions(subscriptions);
    expect(message.toJSON()).toMatchSnapshot();
  });
  test("works for one configuration in a channel", async () => {
    const subscriptions = [{
      authorSlackId: "U12345",
      repoFullName: "atom/atom",
    }];
    const message = new ActivateLegacySubscriptions(subscriptions);
    expect(message.toJSON()).toMatchSnapshot();
  });
});
