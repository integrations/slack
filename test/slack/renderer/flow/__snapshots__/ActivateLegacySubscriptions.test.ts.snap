// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlreadySubscribed message rendering works for multiple configurations and authors in a channel 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": ":wave: Looks like you installed the new GitHub integration on this workspace. Existing configurations have been disabled. Run the following commands to re-enable the configurations in this channel.",
    },
    Object {
      "color": "#24292f",
      "fields": Array [
        Object {
          "value": "\`/github subscribe kubernetes/kubernetes\`",
        },
        Object {
          "value": "\`/github subscribe facebook/react\`",
        },
      ],
      "mrkdwn_in": Array [
        "fields",
      ],
      "pretext": "Configurations created by <@U12345>:",
    },
    Object {
      "color": "#24292f",
      "fields": Array [
        Object {
          "value": "\`/github subscribe django/django\`",
        },
        Object {
          "value": "\`/github subscribe microsoft/typescript\`",
        },
      ],
      "mrkdwn_in": Array [
        "fields",
      ],
      "pretext": "Configurations created by <@U54321>:",
    },
  ],
  "response_type": "in_channel",
}
`;

exports[`AlreadySubscribed message rendering works for one configuration in a channel 1`] = `
Object {
  "attachments": Array [
    Object {
      "color": "#24292f",
      "text": ":wave: Looks like you installed the new GitHub integration on this workspace. Existing configurations have been disabled. Run the following command to re-enable the configuration in this channel.",
    },
    Object {
      "color": "#24292f",
      "fields": Array [
        Object {
          "value": "\`/github subscribe atom/atom\`",
        },
      ],
      "mrkdwn_in": Array [
        "fields",
      ],
      "pretext": "Configuration created by <@U12345>:",
    },
  ],
  "response_type": "in_channel",
}
`;
